src/dataflow/operators/notificator.rs

```diff
@@ -0,0 +1,136 @@
+    pending: Vec<Capability<T>>,
+    frontier: Vec<MutableAntichain<T>>,
+    available: VecDeque<(Capability<T>, i64)>,
+}
+
+impl<T: Timestamp> Notificator<T> {
+    pub fn new() -> Notificator<T> {
+        Notificator {
+            pending: Vec::new(),
+            frontier: Vec::new(),
+            available: VecDeque::new(),
+        }
+    }
+
+    /// Updates the `Notificator`'s frontiers from a `CountMap` per input.
+    pub fn update_frontier_from_cm(&mut self, count_map: &mut [CountMap<T>]) {
+        while self.frontier.len() < count_map.len() {
+            self.frontier.push(MutableAntichain::new());
+        }
+
+        for (index, counts) in count_map.iter_mut().enumerate() {
+            while let Some((time, delta)) = counts.pop() {
+                self.frontier[index].update(&time, delta);
+            }
+        }
+    }
+
+    /// Reveals the elements in the frontier of the indicated input.
+    pub fn frontier(&self, input: usize) -> &[T] {
+        self.frontier[input].elements()
+    }
+
+    /// Requests a notification at the time associated with capability `cap`. Takes ownership of
+    /// the capability.
+    ///
+    /// In order to request a notification at future timestamp, obtain a capability for the new
+    /// timestamp first, as show in the example.
+    ///
+    /// #Examples
+    /// ```
+    /// use timely::dataflow::operators::{ToStream, Unary};
+    /// use timely::dataflow::channels::pact::Pipeline;
+    ///
+    /// timely::example(|scope| {
+    ///     (0..10).to_stream(scope)
+    ///            .unary_notify(Pipeline, "example", Vec::new(), |input, output, notificator| {
+    ///                while let Some((cap, data)) = input.next() {
+    ///                    output.session(&cap).give_content(data);
+    ///                    let mut time = cap.time();
+    ///                    time.inner += 1;
+    ///                    notificator.notify_at(cap.delayed(&time));
+    ///                }
+    ///                while let Some((cap, count)) = notificator.next() {
+    ///                    println!("done with time: {:?}", cap.time());
+    ///                }
+    ///            });
+    /// });
+    /// ```
+    #[inline]
+    pub fn notify_at(&mut self, cap: Capability<T>) {
+        self.pending.push(cap);
+    }
+
+    /// Repeatedly calls `logic` till exhaustion of the available notifications.
+    ///
+    /// `logic` receives a capability for `t`, the timestamp being notified and a `count`
+    /// representing how many capabilities were requested for that specific timestamp.
+    #[inline]
+    pub fn for_each<F: FnMut(Capability<T>, i64)>(&mut self, mut logic: F) {
+        while let Some((cap, count)) = self.next() {
+            ::logging::log(&::logging::GUARDED_PROGRESS, true);
+            logic(cap, count);
+            ::logging::log(&::logging::GUARDED_PROGRESS, false);
+        }
+    }
+
+    #[inline]
+    fn scan_for_available_notifications(&mut self) {
```

Comment:
That might work! I don't read functional code; lemme try google translate on it. Would your life be easier if notifications didn't have a multiplicity, and we just de-dup on insert?

Whilst yoga-ing I was supposed to be thinking of birds and shit, but instead was thinking about how in the fullness of time we probably want the notificator to be throughput optimized (differential wants multiple concurrent notifications, and then to sort work by key rather than by time). This probably means "serve all at once". Just getting that out there.

Queue was *literally* the first timely operator I ever wrote (I think)! It was just supposed to batch things up, not deliver them in any particular order. Actually, it was just supposed to stop an infinite delivery loop. Oops.